logstreams/src/test/java/io/zeebe/distributedlog/restore/impl/DefaultLogReplicationServerHandlerTest.java

```diff
@@ -0,0 +1,225 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.zeebe.distributedlog.restore.impl;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.zeebe.distributedlog.restore.LogReplicationRequest;
+import io.zeebe.distributedlog.restore.LogReplicationResponse;
+import io.zeebe.logstreams.impl.LoggedEventImpl;
+import io.zeebe.logstreams.log.LoggedEvent;
+import io.zeebe.logstreams.util.LogStreamReaderRule;
+import io.zeebe.logstreams.util.LogStreamRule;
+import io.zeebe.logstreams.util.LogStreamWriterRule;
+import io.zeebe.test.util.MsgPackUtil;
+import java.util.ArrayList;
+import java.util.List;
+import org.agrona.DirectBuffer;
+import org.agrona.MutableDirectBuffer;
+import org.agrona.concurrent.UnsafeBuffer;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.RuleChain;
+import org.junit.rules.TemporaryFolder;
+
+public class DefaultLogReplicationServerHandlerTest {
+  private final TemporaryFolder logFolder = new TemporaryFolder();
+  private final LogStreamRule logStreamRule = new LogStreamRule(logFolder);
+  private final LogStreamWriterRule logStreamWriterRule = new LogStreamWriterRule(logStreamRule);
+  private final LogStreamReaderRule logStreamReaderRule = new LogStreamReaderRule(logStreamRule);
+
+  @Rule
+  public RuleChain ruleChain =
+      RuleChain.outerRule(logFolder)
+          .around(logStreamRule)
+          .around(logStreamWriterRule)
+          .around(logStreamReaderRule);
+
+  @Test
+  public void shouldReplicateRequestedEvents() {
+    // given
+    final List<LoggedEvent> loggedEvents = writeEvents(10);
+    final int requestedCount = 10;
+    final int bufferSize = calculateMinimumHandlerBufferSizeForEvents(loggedEvents);
+    final LogReplicationRequest request =
+        new DefaultLogReplicationRequest(
+            loggedEvents.get(0).getPosition(), loggedEvents.get(requestedCount - 1).getPosition());
+    final DefaultLogReplicationServerHandler handler =
+        new DefaultLogReplicationServerHandler(logStreamRule.getLogStream(), bufferSize);
+
+    // when
+    final LogReplicationResponse response = handler.onReplicationRequest(request);
+
+    // then
+    assertThat(response.getToPosition())
+        .isEqualTo(loggedEvents.get(requestedCount - 1).getPosition());
+    assertThat(response.hasMoreAvailable()).isFalse();
+    assertThat(response.getSerializedEvents()).isEqualTo(serialize(loggedEvents));
+    assertThat(response.isValid()).isTrue();
+  }
+
+  @Test
+  public void shouldOnlyReplicateAsMuchAsFitsTheBuffer() {
+    // given
+    final List<LoggedEvent> loggedEvents = writeEvents(10);
+    final int requestedCount = 5;
+    final List<LoggedEvent> expected = loggedEvents.subList(0, requestedCount);
+    final int bufferSize = calculateMinimumHandlerBufferSizeForEvents(expected);
```

Comment:
`onReplicationRequest` cannot handle the case if bufferSize is not the exact size for fitting complete events. You can test it here by setting `bufferSize = calculateMinimumHandlerBufferSizeForEvents(expected) + 10`, for example.

Reply:
Ah, yeah, good catch.